### PR TITLE
Support COPY TO statement

### PR DESCRIPTION
**Pull request by rfdavid**
This PR introduces the command `COPY ([QUERY]) TO [file]`.
The command above will execute the passed query and store the results into the defined file. In this PR, it only supports `.csv` extension, so `.parquet` files will be supported in a separated one.

Example:

```
COPY (MATCH (p:person) RETURN id(p), p.ID, p.fName, p.gender, p.isStudent, p.height) TO "person.csv";

COPY (MATCH (a)<-[:knows|:studyAt]-(b:person) RETURN DISTINCT b.fName) to 'names.csv';
```

Related to #1692

### CODE REVIEW

File: src/include/common/copier_config/copier_config.h
Comment (andyfengHKU) [inline src/include/common/copier_config/copier_config.h:34]: Let's call this CopyDirection to be more clear. Don't forget to change

File: src/parser/transformer.cpp
Comment (andyfengHKU) [inline src/parser/transformer.cpp:1065]: You probably don't need to check direction because the grammar only allows COPY FROM. So give Direction::FROM directly to the constructor

**Review by andyfengHKU — APPROVED** [review]

File: src/binder/bind/bind_copy.cpp
Comment (andyfengHKU) [inline src/binder/bind/bind_copy.cpp:47]: I think we can have following design for `BoundCopy`. Let's start with the simplified version of only CopyTo node table

```
struct BoundCopyFromInfo {
     string tableName;
     share_ptr<NodeExpression> node
}

struct BoundCopyToInfo {
     share_ptr<NodeOrRelExpression> node
     expression_vector expressionsToCopy
}

class BoundCopy {

private:
    CopyDescription
    unique_ptr<BoundCopyFromInfo>
    unique_ptr<BoundCopyToInfo>
}
```

So you need to create a node Expression through `binder.createQueryNode()` and loop through all node properties through `node.getPropertyExpressions` to construct a `expressionsToCopy`

File: src/planner/planner.cpp
Comment (andyfengHKU) [inline src/planner/planner.cpp:193]: If your `boundCopyStatement` has a `nodeExpression`, you can call `joinOrderEnumerator.appendScanNodeID` as the first operator and `queryPlanner.appendScanNodePropIfNecessary` as the second operator. And finally a `CopyTo` on top of it.

File: src/include/processor/operator/copy_to/copy_to.h
Comment (andyfengHKU) [inline src/include/processor/operator/copy_to/copy_to.h:31]: Apart from copyDescription, you also need what vectors are you going to copy to. Which means you need 2 things

- vector<DataPos> vectorsToCopyPos
   -  You should get this from the constructor 
- vector<ValueVector> vectorsToCopy
   -  Initialize this through initLocalStateInternal

You can take a look at `payloads` in `ResultCollector` as a reference

File: src/include/processor/operator/copy_to/copy_to.h
Comment (andyfengHKU) [inline src/include/processor/operator/copy_to/copy_to.h:18]: Here you want do the following

```
if (!children[0].getNext(context)) {
    return false; // No more tuples to pull from child
}
// Here you need to implement the read from vector and copy to csv logic.
```

File: src/processor/mapper/map_ddl.cpp
Comment (andyfengHKU) [inline src/processor/mapper/map_ddl.cpp:62]: The goal of mapper is to construct `vector<DataPos> vectorsToCopyPos`. You can do so by 
```
auto childSchema = logicalOperator.getChild(0).getSchema();
vector<DataPos> vectorsToCopyPos;
for (auto& expression : boundCopy.getCopyToInfo.expressionsToCopy ) {
     vectorsToCopyPos.emplace_back(childSchema.getExpressionPos(expression))
}
```

File: src/binder/bind/bind_copy.cpp
Comment (rfdavid) [reply src/binder/bind/bind_copy.cpp:47]: What do you think about these changes?
1. Move `CopyDirection` to `BoundCopy`
2. Add `CopyDescriptionFrom` and `CopyDescriptionTo` to separate the logic of each one. For example, the current `CopyDescription` contains `CSVReaderConfig`, which is not related to `COPY TO`, and also `filePaths` is a vector, which makes sense for `COPY FROM` but not for `COPY TO`. Inside `CopyDescriptionXXX` (or maybe call `BoundCopyXXXInfo`?) we add the specific properties for each one.

Comment (andyfengHKU) [reply src/binder/bind/bind_copy.cpp]: This design looks good to me.

File: src/processor/processor.cpp
Comment (andyfengHKU) [inline src/processor/processor.cpp:78]: move this line under all `updates` and before `STANDALONE_CALL`. So that we have operators of the same type written together

File: src/antlr4/Cypher.g4
Comment (andyfengHKU) [inline src/antlr4/Cypher.g4:20]: We probably need to separate `COPY FROM\TO` grammar here. Something like

```
kU_CopyFROM : COPY SP oC_SchemaName SP FROM SP kU_FilePaths ( SP? '(' SP? kU_ParsingOptions SP? ')' )? ;

kU_CopyTO: Copy SP oC_Query SP TO kU_FilePaths (eventually there should also be an option here, though might be different from parsing option)
```

File: src/include/binder/copy/bound_copy.h
Comment (andyfengHKU) [inline src/include/binder/copy/bound_copy.h:22]: Split BoundCopy to BoundCopyFrom and BoundCopyTo. BoundCopyFrom should stay the same as it used to be.

BoundCopyTo should have
```
common::CopyDescription copyDescription;
std::unique_ptr<BoundStatement> regularQuery; // or just call it statement
```

File: src/common/copier_config/copier_config.cpp
Comment (andyfengHKU) [inline src/common/copier_config/copier_config.cpp:12]: Once you split BoundCopyTo&From you probably don't need direction any more.

File: src/binder/bind/bind_copy.cpp
Comment (andyfengHKU) [inline src/binder/bind/bind_copy.cpp:16]: once you changed the grammar simply call
```
auto boundFilePaths = copyStatement.getFilePaths();
auto actualFileType = bindFileType(boundFilePaths);
auto boundQuery = bindQuery(copyStatement.getRegularQuery());
return std::make_unique<BoundCopyTo>(std::move(boundQuery),
         CopyDescription(boundFilePaths, actualFileType, CSVReaderConfig{}));
```

File: test/test_files/exceptions/binder/binder_error.test
Comment (andyfengHKU) [inline test/test_files/exceptions/binder/binder_error.test:451]: Don't forget to add a test to check copy to result eventually.

File: src/include/parser/copy.h
Comment (andyfengHKU) [inline src/include/parser/copy.h:12]: I'm not sure if we should split copy to and from here. This is your call.

File: src/include/planner/logical_plan/logical_operator/logical_copy_to.h
Comment (andyfengHKU) [inline src/include/planner/logical_plan/logical_operator/logical_copy_to.h:34]: remove. You should get this from child schema

File: src/include/planner/logical_plan/logical_operator/logical_copy_to.h
Comment (andyfengHKU) [inline src/include/planner/logical_plan/logical_operator/logical_copy_to.h:11]: remove empty line

File: src/planner/planner.cpp
Comment (andyfengHKU) [inline src/planner/planner.cpp:207]: once you have boundStatement in BoundCopyTo. Simply call
```
auto plan = getBestPlan(..., statement)
// append logical copyTo on top
```

File: src/processor/mapper/plan_mapper.cpp
Comment (andyfengHKU) [inline src/processor/mapper/plan_mapper.cpp:211]: You can leave this for now. Let's revisit this part before merging.

File: src/include/parser/copy.h
Comment (andyfengHKU) [inline src/include/parser/copy.h:48]: I guess you only know file type after binding. And it seems this field is not being used. So remove it.

File: src/include/parser/copy.h
Comment (andyfengHKU) [inline src/include/parser/copy.h:49]: Ditto. Seems that tableName is no longer needed.

File: src/include/binder/copy/bound_copy.h
Comment (andyfengHKU) [inline src/include/binder/copy/bound_copy.h:17]: Since you have CopyTo and CopyFrom in the parser part. Let's use BoundCopyFrom and BoundCopyTo for consistency.

File: src/include/common/copier_config/copier_config.h
Comment (andyfengHKU) [inline src/include/common/copier_config/copier_config.h:61]: I might be wrong, but seems you don't need to store copyDirection.

Also I would prefer to merge `copyFromFilePaths` and `copyToFilePath`. In the copyTo case, the vector should be of size 1.

File: src/include/processor/mapper/plan_mapper.h
Comment (andyfengHKU) [inline src/include/processor/mapper/plan_mapper.h:5]: use planner::LogicalOperator rather than planner::LogicalCopy and move this include to cpp file.

File: src/planner/planner.cpp
Comment (andyfengHKU) [inline src/planner/planner.cpp:199]: I think the parser has the guarantee for this already. So you can assert here rather than check and throw

File: src/processor/mapper/plan_mapper.cpp
Comment (andyfengHKU) [inline src/processor/mapper/plan_mapper.cpp:215]: Seems you have figured out the way to return a column for CopyTo. So you can just remove this commented code.

**Review by andyfengHKU — COMMENTED** [review]
Up to planning looks good to me. Good job.

File: src/processor/operator/copy_to/csv_file_writer.cpp
Comment (ray6080) [inline src/processor/operator/copy_to/csv_file_writer.cpp:103]: As discussed, we should assume all flat input for now. So I don't the logic should be complicated as the one you have.
Generally, it should follow something like this:
```
func writeTuple() {
    for (auto vector : vectors) {
        writeValue(vector)
    }
}

func writeValue(vector) {
    assert(vector is flat)
    write value at flatten pos. selectedPos[0]
}
```

**Review by ray6080 — CHANGES_REQUESTED** [review]

File: src/include/binder/copy/bound_copy_to.h
Comment (andyfengHKU) [inline src/include/binder/copy/bound_copy_to.h:32]: tableID and tableName no longer seem to be needed

File: src/include/binder/copy/bound_copy_to.h
Comment (andyfengHKU) [inline src/include/binder/copy/bound_copy_to.h:9]: doesn't seem to be needed

Comment (rfdavid) [reply src/include/binder/copy/bound_copy_to.h:9]: That's right, only `bound_regular_query.h` is needed. I'm removing the others

File: src/common/type_utils.cpp
Comment (rfdavid) [reply src/common/type_utils.cpp:76]: I think we can also speed up with `ostringstream`

**Review by ray6080 — APPROVED** [review]
LGTM!

**Pull request merged**
